// reducers/basket.js
const ADD_PRODUCT = 'ADD_PRODUCT'
const CHANGE_VISIBLE = 'CHANGE_VISIBLE'

let initialState = {
    products: [],
    visible: false
}

export const basket = (state = initialState, {type, payload}) => {
    switch (type) {
        case ADD_PRODUCT:
            let indexOf = state.products.findIndex(item => item.id === payload)
            if (indexOf === -1) {
                return {
                    ...state,
                    products: [...state.products, {id: payload, count: 1}]
                }
            } else {
                const products = [...state.products]
                products[indexOf].count += 1
                return {
                    ...state,
                    products
                }
            }
        case CHANGE_VISIBLE:
            return {
                ...state,
                visible: !state.visible
            }
        default:
            return state
    }
}

export const addProduct = id => dispatch => {
    dispatch({
        type: ADD_PRODUCT,
        payload: id
    })
}

export const changeVisible = () => dispatch => {
    dispatch({
        type: CHANGE_VISIBLE
    })
}
