// reducers/basket.js
const ADD_PRODUCT = 'ADD_PRODUCT'
const CHANGE_VISIBLE = 'CHANGE_VISIBLE'
const DELETE_FROM_BASKET = 'DELETE_FROM_BASKET'
const DELETE_PRODUCT = 'DELETE_PRODUCT'

let initialState = {
    products: [],
    visible: false
}

export const basket = (state = initialState, {type, payload}) => {
    switch (type) {
        case ADD_PRODUCT:
            let indexOf = state.products.findIndex(item => item.id === payload)
            if (indexOf === -1) {
                return {
                    ...state,
                    products: [...state.products, {id: payload, count: 1}]
                }
            } else {
                const products = [...state.products]
                products[indexOf].count += 1
                return {
                    ...state,
                    products
                }
            }
        case CHANGE_VISIBLE:
            return {
                ...state,
                visible: !state.visible
            }
        case DELETE_FROM_BASKET:
            return {
                ...state,
                products: state.products.filter(product => product.id != payload)
            }
        case DELETE_PRODUCT:
            const itemIndex = state.products.findIndex(item => item.id === payload),
                products = [...state.products]
            products[itemIndex].count -= 1
            if (products[itemIndex].count === 0) {
                return {
                    ...state,
                    products: state.products.filter(item => item.id !== payload)
                }
            }
            return {
                ...state,
                products,
            }
        default:
            return state
    }
}

export const addProduct = id => dispatch => {
    dispatch({
        type: ADD_PRODUCT,
        payload: id
    })
}

export const changeVisible = () => dispatch => {
    dispatch({
        type: CHANGE_VISIBLE
    })
}

export const deleteFromBasket = id => dispatch => {
    dispatch({
        type: DELETE_FROM_BASKET,
        payload: id
    })
}

export const deleteProduct = id => dispatch => {
    dispatch({
        type: DELETE_PRODUCT,
        payload: id
    })
}
