// reducers/shop.js
const GET_ITEMS_START = 'GET_ITEMS_START'
const GET_ITEMS_SUCCESS = 'GET_ITEMS_SUCCESS'
const GET_ITEMS_FAILURE = 'GET_ITEMS_FAILURE'

import {getItems as getItemsApi} from 'api'

let initialState = {
    items: [],
    loading: false
}

export const shop = (state = initialState, {type, payload}) => {
    switch (type) {
        case GET_ITEMS_START:
            return {
                ...state,
                loading: true
            }
        case GET_ITEMS_SUCCESS:
            return {
                ...state,
                items: payload.map(item => ({...item, count: 0})),
                loading: false
            }
        default:
            return state
    }
}

export const getItems = () => async dispatch => {
    dispatch({
        type: GET_ITEMS_START
    })
    try {
        const items = await getItemsApi()
        dispatch({
            type: GET_ITEMS_SUCCESS,
            payload: items
        })
    } catch (err) {
        dispatch({
            type: GET_ITEMS_FAILURE,
            payload: err,
            error: true
        })
    }
}
